src/mpid/ch4/shm/xpmem/xpmem_control.h

```diff
@@ -99,15 +79,52 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr
 
     if (rreq) {
         /* Matching receive was posted */
-        MPIR_Comm_release(root_comm);   /* -1 for posted_list */
         MPIDIG_REQUEST(rreq, rank) = slmt_req_hdr->src_rank;
         MPIDIG_REQUEST(rreq, tag) = slmt_req_hdr->tag;
         MPIDIG_REQUEST(rreq, context_id) = slmt_req_hdr->context_id;
 
         /* Complete XPMEM receive */
-        mpi_errno = MPIDI_XPMEM_handle_lmt_recv(slmt_req_hdr->src_offset, slmt_req_hdr->data_sz,
-                                                slmt_req_hdr->sreq_ptr, slmt_req_hdr->src_lrank,
-                                                rreq);
+#ifdef MPIDI_CH4_SHM_XPMEM_COOP_P2P
+        int recvtype_iscontig;
+        MPIDI_av_entry_t *av;
+        /* TODO: need to support non-contig datatype on receiver side. */
+        MPIR_Datatype_iscontig(MPIDIG_REQUEST(rreq, datatype), &recvtype_iscontig);
+        if (recvtype_iscontig) {
+            MPIDI_XPMEM_REQUEST(rreq, counter) =
+                &MPIDI_XPMEM_global.coop_counter[MPIDI_XPMEM_global.local_rank *
+                                                 MPIDI_XPMEM_global.num_local +
+                                                 slmt_req_hdr->src_lrank];
+
+            av = MPIDIU_comm_rank_to_av(root_comm, slmt_req_hdr->src_rank);
+            MPIDI_XPMEM_REQUEST(rreq, sreq_ptr) = slmt_req_hdr->sreq_ptr;
+
+            mpi_errno = MPIDI_XPMEM_lmt_coop_isend(MPIDIG_REQUEST(rreq, buffer),
+                                                   MPIDIG_REQUEST(rreq, count),
+                                                   MPIDIG_REQUEST(rreq, datatype),
+                                                   slmt_req_hdr->src_rank, slmt_req_hdr->tag,
+                                                   root_comm, MPIR_CONTEXT_INTRA_PT2PT, av,
+                                                   MPIDI_SHM_XPMEM_SEND_LMT_CTS, &rreq);
+            if (MPI_SUCCESS != mpi_errno)
+                MPIR_ERR_POP(mpi_errno);
+        } else {
+            MPIDI_XPMEM_REQUEST(rreq, counter) = NULL;
+        }
+
+        mpi_errno =
+            MPIDI_XPMEM_handle_lmt_coop_recv(slmt_req_hdr->src_offset, slmt_req_hdr->data_sz,
+                                             slmt_req_hdr->sreq_ptr, slmt_req_hdr->src_lrank,
+                                             MPIDI_SHM_XPMEM_SEND_LMT_RTS, rreq);
+        if (MPI_SUCCESS != mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+#else
+        mpi_errno =
+            MPIDI_XPMEM_handle_lmt_single_recv(slmt_req_hdr->src_offset, slmt_req_hdr->data_sz,
```

Comment:
As previously mentioned, I don't like having a bunch of receive functions. I'd rather see this as a single receive function where it can internally determine how many ways to chunk the message (where the number of chunks could be `1` for no chunking / cooperation.

Reply:
Our previous idea is keeping receive function constant in shm layer and in xpmem layer we have a switch (do single copy or coop copy). Current implementation sustains single XPMEM copy code so I still keep it there. If you think it is redundant, I need to delete single side code and merge them into one.

Reply:
Yes, I would rather see the existing single copy code go away *if* the cooperative copy code is always at least as good. It may be that there are cases where the cooperative copy isn't better, but as long as it's not worse, there's no reason to keep the old code around.

Reply:
So do we need just delete single copy code and all related data definition in this PR and only keep coop P2P code?

Reply:
I think we should delete it (it’s Git so we can always find it later). You’re right that we don’t need the CVAR anymore.